File: weblayer/browser/browser_list.cc
// Copyright 2020 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "weblayer/browser/browser_list.h"

#include <functional>

#include "base/no_destructor.h"
#include "base/ranges/algorithm.h"
#include "build/build_config.h"
#include "weblayer/browser/browser_impl.h"
#include "weblayer/browser/browser_list_observer.h"

#include "weblayer/browser/browser_list_proxy.h"

namespace weblayer {

// static
BrowserList* BrowserList::GetInstance() {
  static base::NoDestructor<BrowserList> browser_list;
  return browser_list.get();
}

void BrowserList::AddObserver(BrowserListObserver* observer) {
  observers_.AddObserver(observer);
}

void BrowserList::RemoveObserver(BrowserListObserver* observer) {
  observers_.RemoveObserver(observer);
}

BrowserList::BrowserList() {
  browser_list_proxy_ = std::make_unique<BrowserListProxy>();
  AddObserver(browser_list_proxy_.get());
}

BrowserList::~BrowserList() {
  RemoveObserver(browser_list_proxy_.get());
}

void BrowserList::AddBrowser(BrowserImpl* browser) {
  DCHECK(!browsers_.contains(browser));
  browsers_.insert(browser);
  for (BrowserListObserver& observer : observers_)
    observer.OnBrowserCreated(browser);
}

void BrowserList::RemoveBrowser(BrowserImpl* browser) {
  DCHECK(browsers_.contains(browser));
  browsers_.erase(browser);

  for (BrowserListObserver& observer : observers_)
    observer.OnBrowserDestroyed(browser);
}

}  // namespace weblayer

File: weblayer/browser/java/org/chromium/weblayer_private/BrowserList.java
// Copyright 2020 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.weblayer_private;

import androidx.annotation.NonNull;

import org.chromium.base.ObserverList;
import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;

/**
 * Tracks the set of Browsers.
 */
@JNINamespace("weblayer")
public final class BrowserList {
    private static BrowserList sInstance;
    private final ObserverList<BrowserListObserver> mObservers;

    @CalledByNative
    private static BrowserList createBrowserList() {
        // The native side should call this only once.
        assert sInstance == null;
        sInstance = new BrowserList();
        return sInstance;
    }

    @NonNull
    public static BrowserList getInstance() {
        // The native side creates this early on. It should never be null.
        if (sInstance == null) {
            BrowserListJni.get().createBrowserList();
            assert sInstance != null;
        }
        return sInstance;
    }

    private BrowserList() {
        mObservers = new ObserverList<>();
    }

    public void addObserver(BrowserListObserver o) {
        mObservers.addObserver(o);
    }

    public void removeObserver(BrowserListObserver o) {
        mObservers.removeObserver(o);
    }

    @CalledByNative
    private void onBrowserCreated(BrowserImpl browser) {
        for (BrowserListObserver observer : mObservers) {
            observer.onBrowserCreated(browser);
        }
    }

    @CalledByNative
    private void onBrowserDestroyed(BrowserImpl browser) {
        for (BrowserListObserver observer : mObservers) {
            observer.onBrowserDestroyed(browser);
        }
    }

    @NativeMethods
    interface Natives {
        void createBrowserList();
    }
}
